### PR TITLE
CrudService

### PR DESCRIPTION
**Pull request by HerbertBodner**
Generic CrudService implementation (without CrudController)

To keep the PR "small", following functionality is NOT part of it:
 * Validation in Service Layer using Validator interfaces
 * Extension method RepositoryConfiguration.AddCrudRepository and RepositoryConfiguration.AddCrudRepositoriesFromAssemblyOf
#27

### CODE REVIEW

**Comment by HerbertBodner** [comment]
Sorry @rurku @yshchohaleu I'll try to keep my PRs smaller next time

File: src/WaCore.Crud.Contracts/Services/IWacListDataService.cs [hunk 1]
```diff
@@ -8,6 +8,10 @@ namespace WaCore.Crud.Contracts.Services
 {
     public interface IWacListDataService<TFilter, TDto>
     {
+        TDto Get(object id);
+        Task<TDto> GetAsync(object id);
+        IPagedList<TDto> GetAll(TFilter filter);
```
Comment (rurku) [inline src/WaCore.Crud.Contracts/Services/IWacListDataService.cs]: I don't like the name GetAll. If it takes a filter object, it means it doesn't always return ALL records. I'd call it GetList.

File: src/WaCore.Crud.ListSample1/Controllers/CarsController.cs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        }
+
+
+        public virtual async Task<IActionResult> Delete(int id)
+        {
+            try
+            {
+                await _service.DeleteAsync(id);
+
+                SetSuccessMessage("Deleted successful");
```
Comment (rurku) [inline src/WaCore.Crud.ListSample1/Controllers/CarsController.cs]: successfully

File: src/WaCore.Crud.ListSample1/Controllers/CarsController.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            {
+                // ToDo Logging and proper error handling
+                ModelState.AddModelError("Error during Edit", ex.Message);
+
+                return await PrepareEditVmAsync(vm);
+            }
+        }
+
+
+        public virtual async Task<IActionResult> Delete(int id)
```
Comment (rurku) [inline src/WaCore.Crud.ListSample1/Controllers/CarsController.cs]: Delete should be a POST with anti-forgery token

File: src/WaCore.Crud.ListSample1/Data/DbInitializer.cs [hunk 4]
```diff
@@ -18,6 +18,34 @@ public static void Initialize(LibraryDbContext dbContext)
             dbContext.Add(new Book { Title = "Crime and Punishment", Author = "Fyodor Dostoyevsky" });
             dbContext.Add(new Book { Title = "The Odyssey", Author = "Homer" });
             dbContext.SaveChanges();
+
+            
+            dbContext.Add(new Car {Brand = "Audi", Model = "A1", CreatedAt = new DateTime(2001, 1, 1) });
```
Comment (rurku) [inline src/WaCore.Crud.ListSample1/Data/DbInitializer.cs]: I see that you like German cars

File: src/WaCore.Crud.ListSample1/Data/UnitOfWork.cs [hunk 5]
```diff
@@ -14,11 +14,15 @@ public UnitOfWork(LibraryDbContext dbContext, IServiceProvider serviceProvider)
         { }
 
         public IBooksListRepository BooksRepository => GetRepository<IBooksListRepository>();
+
+        public ICarRepository CarRepository => GetRepository<ICarRepository>();
```
Comment (rurku) [inline src/WaCore.Crud.ListSample1/Data/UnitOfWork.cs]: For consistency you should call it CarsRepository or rename the other one to BookRepository

File: src/WaCore.Crud.Contracts/Web/ViewModels/IWacListDataVm.cs [hunk 6]
```diff
@@ -2,14 +2,14 @@
 using System.Collections.Generic;
 using System.Text;
 using WaCore.Crud.Contracts.Dtos;
+using WaCore.Crud.Contracts.Utils;
 
-namespace WaCore.Crud.Contracts.Web
+namespace WaCore.Crud.Contracts.Web.ViewModels
 {
     public interface IWacListDataVm<TDto, TFilter>
```
Comment (rurku) [inline src/WaCore.Crud.Contracts/Web/ViewModels/IWacListDataVm.cs]: This interface is not used

File: src/WaCore.Crud.Contracts/Web/ViewModels/IWacEditVm.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace WaCore.Crud.Contracts.Web.ViewModels
+{
+    public interface IWacEditVm<TDto, TNewDto>
```
Comment (rurku) [inline src/WaCore.Crud.Contracts/Web/ViewModels/IWacEditVm.cs]: This interface is not used

File: src/WaCore.Crud.ListSample1/Views/Home/Index.cshtml [hunk 8]
```diff
@@ -18,5 +18,17 @@
         <div>@Html.ActionLink("View example", "Index", "BookListData") </div>
 
     </div>
+
+    <div class="col-md-6">
+        <h2>Cars list</h2>
+        <h4>Simple CRUD functionality</h4>
+        <div>
+            Use base classes for Service (WacCrudService), Controller (WacCrudController), Filter (WacFilter), List-Viewmodel (WacListDataVm) and Edit-Viewmodel (WacEditVm).
```
Comment (rurku) [inline src/WaCore.Crud.ListSample1/Views/Home/Index.cshtml]: You don't have WacCrudController, WacListDataVm and WacEditVm.
And you forgot to mention WacListDataRepository.

File: src/WaCore.Crud.Utils/Exceptions/ResourceNotFoundException.cs [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+        public ResourceNotFoundException(object resource, string message)
+            : base(message)
+        {
+            _resource = resource;
+        }
+
+        public object Resource { get { return _resource; } }
+
+        public string BaseMessage
```
Comment (rurku) [inline src/WaCore.Crud.Utils/Exceptions/ResourceNotFoundException.cs]: You don't use BaseMessage

File: src/webapp-core.sln [hunk 10]
```diff
@@ -33,9 +31,9 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Tests", "Tests", "{CBB0AA48
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "WaCore.Crud.UnitTests", "WaCore.Crud.UnitTests\WaCore.Crud.UnitTests.csproj", "{13973730-A24F-4059-81C9-BFE300EC406D}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "WaCore.Web", "WaCore.Web\WaCore.Web.csproj", "{6BDA7B6E-2853-415E-9C27-EAE603AAB7BD}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "WaCore.Web", "WaCore.Web\WaCore.Web.csproj", "{6BDA7B6E-2853-415E-9C27-EAE603AAB7BD}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "WaCore.Sample.Middlewares", "WaCore.Sample.Middlewares\WaCore.Sample.Middlewares.csproj", "{504C2CFA-A7D3-4D11-A1B9-0CB4D78521E0}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "WaCore.Sample.Middlewares", "WaCore.Sample.Middlewares\WaCore.Sample.Middlewares.csproj", "{504C2CFA-A7D3-4D11-A1B9-0CB4D78521E0}"
```
Comment (rurku) [inline src/webapp-core.sln]: I know it's not a part of this pull request, but the word 'middleware' is uncountable, so it does not have a plural form. Just like software, hardware and firmware.

File: src/WaCore.Crud.Contracts/Services/IWacListDataService.cs (hunk 1 above)
Comment (HerbertBodner) [reply src/WaCore.Crud.Contracts/Services/IWacListDataService.cs]: Renamed to GetList

File: src/WaCore.Crud.ListSample1/Controllers/CarsController.cs (hunk 2 above)
Comment (HerbertBodner) [reply src/WaCore.Crud.ListSample1/Controllers/CarsController.cs]: Done

File: src/WaCore.Crud.ListSample1/Controllers/CarsController.cs (hunk 3 above)
Comment (HerbertBodner) [reply src/WaCore.Crud.ListSample1/Controllers/CarsController.cs]: Done, but it is not used yet. I created following issue to implement it: #45

File: src/WaCore.Crud.ListSample1/Data/UnitOfWork.cs (hunk 5 above)
Comment (HerbertBodner) [reply src/WaCore.Crud.ListSample1/Data/UnitOfWork.cs]: Renamed the IBookListRepository

File: src/WaCore.Crud.Contracts/Web/ViewModels/IWacListDataVm.cs (hunk 6 above)
Comment (HerbertBodner) [reply src/WaCore.Crud.Contracts/Web/ViewModels/IWacListDataVm.cs]: Deleted

File: src/WaCore.Crud.Contracts/Web/ViewModels/IWacEditVm.cs (hunk 7 above)
Comment (HerbertBodner) [reply src/WaCore.Crud.Contracts/Web/ViewModels/IWacEditVm.cs]: Deleted

File: src/WaCore.Crud.ListSample1/Views/Home/Index.cshtml (hunk 8 above)
Comment (HerbertBodner) [reply src/WaCore.Crud.ListSample1/Views/Home/Index.cshtml]: Corrected

File: src/WaCore.Crud.Utils/Exceptions/ResourceNotFoundException.cs (hunk 9 above)
Comment (HerbertBodner) [reply src/WaCore.Crud.Utils/Exceptions/ResourceNotFoundException.cs]: Removed

File: src/webapp-core.sln (hunk 10 above)
Comment (HerbertBodner) [reply src/webapp-core.sln]: Created an issue for it: #46

File: src/WaCore.Crud.ListSample1/Data/DbInitializer.cs (hunk 4 above)
Comment (HerbertBodner) [reply src/WaCore.Crud.ListSample1/Data/DbInitializer.cs]: Yes :)

File: src/WaCore.Crud.ListSample1/Views/Home/Index.cshtml [hunk 11]
```diff
@@ -23,8 +30,16 @@
...
         <h4>Simple CRUD functionality</h4>
         <div>
-            Use base classes for Service (WacCrudService), Controller (WacCrudController), Filter (WacFilter), List-Viewmodel (WacListDataVm) and Edit-Viewmodel (WacEditVm).
-            By deriving your own classes from those base classes, you only need to impmlement their abstract methods and get a nicely structured, extendable code, which has CRUD functionality.
+            <p>
+                Same as the books example, but provides also CRUD functionality.
+            </p>
+            <p>
+                It uses the same base classes as the books example for Repository (<code>WacListDataRepository</code>) and Filter (<code>WacFilter</code>).
+                But instead of using  for <code>WacListDataService</code> for the base class of the service, it uses <code>WacCrudService</code>, which extends the former one with CRUD functionality.
```
Comment (rurku) [inline src/WaCore.Crud.ListSample1/Views/Home/Index.cshtml]: One `for` too many

File: src/WaCore.Crud.ListSample1/Data/UnitOfWork.cs (hunk 5 above)
Comment (rurku) [reply src/WaCore.Crud.ListSample1/Data/UnitOfWork.cs]: But the property in UnitOfWork is still called BooksRepository

File: src/WaCore.Crud.ListSample1/Views/Home/Index.cshtml (hunk 11 above)
Comment (HerbertBodner) [reply src/WaCore.Crud.ListSample1/Views/Home/Index.cshtml]: fixed

File: src/WaCore.Crud.ListSample1/Data/UnitOfWork.cs (hunk 5 above)
Comment (HerbertBodner) [reply src/WaCore.Crud.ListSample1/Data/UnitOfWork.cs]: renamed as well

**Pull request merged**